ibind/oauth/oauth2.py

```diff
@@ -0,0 +1,289 @@
+# Combined imports
+from dataclasses import dataclass, field
+from typing import Optional
+from ibind import var
+from ibind.oauth import OAuthConfig # For OAuth2Config parent
+import base64
+import json
+import math
+import pprint
+import requests
+import time
+from Crypto.PublicKey import RSA
+from Crypto.Signature import PKCS1_v1_5
+from Crypto.Hash import SHA256
+from ibind.support.logs import project_logger
+
+# OAuth2Config class definition
+@dataclass
+class OAuth2Config(OAuthConfig):
+    """
+    Dataclass encapsulating OAuth 2.0 configuration parameters.
+    """
+
+    # --- Core OAuth 2.0 Parameters ---
+    client_id: str = var.IBIND_OAUTH2_CLIENT_ID
+    client_key_id: str = var.IBIND_OAUTH2_CLIENT_KEY_ID
+
+    # --- Direct Credential Storage ---
+    private_key_pem: Optional[str] = var.IBIND_OAUTH2_PRIVATE_KEY_PEM
+    username: Optional[str] = var.IBIND_OAUTH2_USERNAME
+
+    # Optional: Pre-configured IP address. If None, will be auto-fetched.
+    ip_address: Optional[str] = var.IBIND_OAUTH2_IP_ADDRESS
+
+    # --- OAuth 2.0 Endpoints and Settings (with defaults) ---
+    token_url: str = field(default=var.IBIND_OAUTH2_TOKEN_URL or 'https://api.ibkr.com/oauth2/api/v1/token')
+    sso_session_url: str = field(default=var.IBIND_OAUTH2_SSO_SESSION_URL or 'https://api.ibkr.com/gw/api/v1/sso-sessions')
+    audience: str = field(default=var.IBIND_OAUTH2_AUDIENCE or '/token')
+    scope: str = field(default=var.IBIND_OAUTH2_SCOPE or 'sso-sessions.write')
+
+    # --- IBKR API Base URL (after successful OAuth) ---
+    oauth_rest_url: str = var.IBIND_OAUTH2_REST_URL or var.IBIND_REST_URL or 'https://api.ibkr.com/v1/api/'
+    oauth_ws_url: str = var.IBIND_OAUTH2_WS_URL or var.IBIND_WS_URL or 'wss://api.ibkr.com/v1/api/ws'
+
+    # --- Token Storage ---
+    access_token: Optional[str] = field(default=None, init=False)
+    sso_bearer_token: Optional[str] = field(default=None, init=False)
+
+    def version(self):
+        return 2.0
+
+    def verify_config(self) -> None:
+        required_params = [
+            'client_id',
+            'client_key_id',
+            'private_key_pem',
+            'username',
+            'token_url',
+            'sso_session_url',
+            'audience',
+            'scope'
+        ]
+        missing_params = [param for param in required_params if not getattr(self, param)]
+        if missing_params:
+            raise ValueError(f'OAuth2Config is missing required parameters: {", ".join(missing_params)}')
+        super().verify_config()
+
+# Original oauth2.py logic starts here
+_LOGGER = project_logger(__file__)
+
+def _get_public_ip():
+    """Fetches the public IP address from an external service."""
+    try:
+        response = requests.get("https://api.ipify.org?format=json", timeout=5)
+        response.raise_for_status()
+        ip_address = response.json()["ip"]
+        _LOGGER.info(f"Successfully fetched public IP address: {ip_address}")
```

Comment:
this looks more like a debug log than info - in normal functionality the user will not need to know this